hal/armv7a/zynq7000/interrupts.c

```diff
@@ -19,52 +19,225 @@
+};
 
 
 struct {
 	volatile u32 *gic;
 	spinlock_t spinlock[SIZE_INTERRUPTS];
 	intr_handler_t *handlers[SIZE_INTERRUPTS];
 	unsigned int counters[SIZE_INTERRUPTS];
-} interrupts;
+} interrupts_common;
 
 
-extern void _end(void);
+/* Required configuration for SPI (Shared Peripheral Interrupts IRQID[32:95]) */
+static const u8 spiConf[] = {/* IRQID: 32-39 */ rising_edge, rising_edge, high_lvl, high_lvl, reserved, high_lvl, high_lvl, high_lvl,
+                             /* IRQID: 40-47 */ high_lvl, rising_edge, high_lvl, high_lvl, high_lvl, high_lvl, high_lvl, high_lvl,
+                             /* IRQID: 48-55 */ high_lvl, high_lvl, high_lvl, high_lvl, high_lvl, high_lvl, high_lvl, rising_edge,
+                             /* IRQID: 56-63 */ high_lvl, high_lvl, high_lvl, high_lvl, high_lvl, rising_edge, rising_edge, rising_edge,
+                             /* IRQID: 64-71 */ rising_edge, rising_edge, rising_edge, rising_edge, rising_edge, high_lvl, high_lvl, high_lvl,
+                             /* IRQID: 72-79 */ high_lvl, high_lvl, high_lvl, high_lvl, high_lvl, high_lvl, rising_edge, high_lvl,
+                             /* IRQID: 80-87 */ high_lvl, high_lvl, high_lvl, high_lvl, high_lvl, high_lvl, high_lvl, high_lvl,
+                             /* IRQID: 88-95 */ high_lvl, high_lvl, high_lvl, high_lvl, rising_edge, reserved, reserved, reserved };
+
+
+
+extern int threads_schedule(unsigned int n, cpu_context_t *context, void *arg);
+
+extern unsigned int _end;
 
 
 void interrupts_dispatch(unsigned int n, cpu_context_t *ctx)
 {
+	intr_handler_t *h;
+	int reschedule = 0;
+	spinlock_ctx_t sc;
+
+	n = *(interrupts_common.gic + ciar) & 0x3ff;
+
+	if (n >= SIZE_INTERRUPTS)
+		return;
+
+	hal_spinlockSet(&interrupts_common.spinlock[n], &sc);
+
+	interrupts_common.counters[n]++;
+
+	if ((h = interrupts_common.handlers[n]) != NULL) {
+		do
+			reschedule |= h->f(n, ctx, h->data);
+		while ((h = h->next) != interrupts_common.handlers[n]);
+	}
+
+	if (reschedule)
+		threads_schedule(n, ctx, NULL);
+
+	*(interrupts_common.gic + ceoir) = (*(interrupts_common.gic + ceoir) & ~0x3ff) | n;
+
+	hal_spinlockClear(&interrupts_common.spinlock[n], &sc);
+
+	return;
+}
+
+
+static void interrupts_enableIRQ(unsigned int irqn)
+{
+	*(interrupts_common.gic + diser0 + (irqn >> 5)) = 1 << (irqn & 0x1f);
+}
+
+
+static void interrupts_disableIRQ(unsigned int irqn)
+{
+	*(interrupts_common.gic + dicer0 + (irqn >> 5)) = 1 << (irqn & 0x1f);
+}
+
+
+static void interrupts_setConf(unsigned int irqn, u32 conf)
+{
+	u32 mask;
+
+	if (irqn < SPI_FIRST_IRQID || irqn >= SIZE_INTERRUPTS || conf == reserved)
+		return;
+
+	mask = *(interrupts_common.gic + dicfr0 + (irqn >> 4)) & ~(0x3 << ((irqn & 0xf) << 1));
+	*(interrupts_common.gic + dicfr0 + (irqn >> 4)) = mask | ((conf & 0x3) << ((irqn & 0xf) << 1));
+}
+
+
+static void interrupts_setCPU(unsigned int irqn, u32 cpuID)
+{
+	u32 mask;
+
+	if (irqn < SPI_FIRST_IRQID || irqn >= SIZE_INTERRUPTS)
+		return;
```

Comment:
Some functions have this sanity check, some not. You're using this functions only within this file in a controlled manner, so I think there is no need for this extra defensiveness

Reply:
Applies to `interrupts_setConf` too

Reply:
Done